File: NoriTests/HistoryStoreTests.swift
import Foundation
import SwiftData
import Testing
@testable import Nori

@Suite("HistoryStore", .serialized)
@MainActor
struct HistoryStoreTests {
    /// SwiftData traps when several containers for the same models live in one process,
    /// so tests share the app's in-memory container and start from an empty store.
    private func makeStore() throws -> HistoryStore {
        let store = HistoryStore(context: Storage.shared.context)
        store.load()
        store.clear(includingPinned: true)
        return store
    }

    private func textDraft(_ text: String, app: String = "com.apple.Notes") -> ClipDraft {
        var draft = ClipClassifier.makeDraft(contents: [.init(type: PasteboardType.utf8PlainText, data: Data(text.utf8))])!
        draft.sourceBundleID = app
        return draft
    }

    @Test func ingestOrdersNewestFirst() throws {
        let store = try makeStore()
        store.ingest(textDraft("one"), now: Date(timeIntervalSince1970: 1))
        store.ingest(textDraft("two"), now: Date(timeIntervalSince1970: 2))
        #expect(store.rows.map(\.title) == ["two", "one"])
    }

    @Test func duplicatesBumpInsteadOfInserting() throws {
        let store = try makeStore()
        let first = store.ingest(textDraft("same"), now: Date(timeIntervalSince1970: 1))
        store.ingest(textDraft("other"), now: Date(timeIntervalSince1970: 2))
        let again = store.ingest(textDraft("same", app: "com.apple.Safari"), now: Date(timeIntervalSince1970: 3))
        #expect(again == first)
        #expect(store.rows.count == 2)
        let row = try #require(store.row(id: again))
        #expect(store.rows.first?.id == again)
        #expect(row.copyCount == 2)
        #expect(row.sourceBundleID == "com.apple.Safari")
        #expect(row.firstCopiedAt == Date(timeIntervalSince1970: 1))
        #expect(row.lastCopiedAt == Date(timeIntervalSince1970: 3))
    }

    @Test func limitDropsOldestUnpinned() throws {
        let store = try makeStore()
        store.maxItems = 3
        for i in 0..<5 {
            store.ingest(textDraft("item \(i)"), now: Date(timeIntervalSince1970: TimeInterval(i)))
        }
        #expect(store.rows.map(\.title) == ["item 4", "item 3", "item 2"])
    }

    @Test func loweringTheLimitPersistsTheEvictions() throws {
        let store = try makeStore()
        for i in 0..<5 {
            store.ingest(textDraft("item \(i)"), now: Date(timeIntervalSince1970: TimeInterval(i)))
        }
        store.maxItems = 2
        #expect(store.rows.map(\.title) == ["item 4", "item 3"])
        // A fresh context sees only what was saved.
        let fresh = ModelContext(Storage.shared.container)
        #expect(try fresh.fetchCount(FetchDescriptor<ClipItem>()) == 2)
    }

    @Test func unpinningAtTheLimitKeepsTheItem() throws {
        let store = try makeStore()
        store.maxItems = 2
        let old = store.ingest(textDraft("old"), now: Date(timeIntervalSince1970: 0))
        store.togglePin(id: old)
        store.ingest(textDraft("a"), now: Date(timeIntervalSince1970: 1))
        store.ingest(textDraft("b"), now: Date(timeIntervalSince1970: 2))
        store.togglePin(id: old)
        #expect(store.row(id: old)?.isPinned == false)
        #expect(store.rows.map(\.title) == ["b", "a", "old"], "the cap applies on the next capture, not now")
        store.ingest(textDraft("c"), now: Date(timeIntervalSince1970: 3))
        #expect(store.rows.map(\.title) == ["c", "b"])
    }

    @Test func restoreMergesIntoARowWithTheSameContent() throws {
        let store = try makeStore()
        let original = store.ingest(textDraft("same"), now: Date(timeIntervalSince1970: 1))
        store.togglePin(id: original, now: Date(timeIntervalSince1970: 2))
        let record = try #require(store.delete(id: original))
        // Re-copied within the undo window: a fresh row exists when ⌘Z arrives.
        let again = store.ingest(textDraft("same"), now: Date(timeIntervalSince1970: 5))
        let restored = store.restore(record)
        #expect(restored == again)
        #expect(store.rows.count == 1)
        let row = try #require(store.row(id: again))
        #expect(row.copyCount == 2)
        #expect(row.firstCopiedAt == Date(timeIntervalSince1970: 1))
        #expect(row.lastCopiedAt == Date(timeIntervalSince1970: 5))
        #expect(row.isPinned)
    }

    @Test func payloadsAreReadOnDemand() throws {
        let store = try makeStore()
        let id = store.ingest(textDraft("payload"), now: Date(timeIntervalSince1970: 1))
        #expect(store.contents(id: id).map(\.type) == [PasteboardType.utf8PlainText])
        #expect(store.plainText(id: id) == "payload")
        #expect(store.imageData(id: id) == nil)
        #expect(store.contents(id: UUID()).isEmpty)
        store.delete(id: id)
        #expect(store.contents(id: id).isEmpty)
        #expect(store.plainText(id: id) == nil)
    }

    @Test func pinnedItemsSurviveLimitAndClear() throws {
        let store = try makeStore()
        store.maxItems = 2
        let keep = store.ingest(textDraft("keep"), now: Date(timeIntervalSince1970: 0))
        store.togglePin(id: keep)
        for i in 1...4 {
            store.ingest(textDraft("item \(i)"), now: Date(timeIntervalSince1970: TimeInterval(i)))
        }
        #expect(store.rows.contains { $0.id == keep })
        #expect(store.rows.filter { !$0.isPinned }.count == 2)

        #expect(store.clear(includingPinned: false) == 2)
        #expect(store.rows.map(\.title) == ["keep"])
        store.clear(includingPinned: true)
        #expect(store.rows.isEmpty)
    }

    @Test func deleteUndoAndPersistence() throws {
        let store = try makeStore()
        let a = store.ingest(textDraft("a"), now: Date(timeIntervalSince1970: 1))
        store.togglePin(id: a, now: Date(timeIntervalSince1970: 5))
        store.ingest(textDraft("b"), now: Date(timeIntervalSince1970: 2))
        let record = try #require(store.delete(id: a))
        #expect(store.rows.map(\.title) == ["b"])
        store.load()
        #expect(store.rows.map(\.title) == ["b"])

        let restored = store.restore(record)
        let row = try #require(store.row(id: restored))
        #expect(row.title == "a")
        #expect(row.isPinned)
        #expect(row.firstCopiedAt == Date(timeIntervalSince1970: 1))
        #expect(store.rows.map(\.title) == ["b", "a"])
    }

    @Test func touchMovesToTop() throws {
        let store = try makeStore()
        let a = store.ingest(textDraft("a"), now: Date(timeIntervalSince1970: 1))
        store.ingest(textDraft("b"), now: Date(timeIntervalSince1970: 2))
        store.touch(id: a, now: Date(timeIntervalSince1970: 3))
        #expect(store.rows.first?.id == a)
        #expect(store.row(id: a)?.copyCount == 2)
    }

    @Test func expiryDropsOldUnpinnedOnly() throws {
        let store = try makeStore()
        store.expireAfterDays = 7
        let now = Date(timeIntervalSince1970: 100 * 86_400)
        let old = store.ingest(textDraft("old"), now: now.addingTimeInterval(-10 * 86_400))
        let oldPinned = store.ingest(textDraft("old pinned"), now: now.addingTimeInterval(-10 * 86_400))
        store.togglePin(id: oldPinned, now: now)
        store.ingest(textDraft("fresh"), now: now.addingTimeInterval(-86_400))
        store.expireOldItems(now: now)
        #expect(store.row(id: old) == nil)
        #expect(store.rows.map(\.title).sorted() == ["fresh", "old pinned"])
    }
}

File: windows/Nori.Core.Tests/HistoryStoreTests.cs
using Nori.Core;
using Xunit;
using static Nori.Core.Tests.TestRows;

namespace Nori.Core.Tests;

public class HistoryStoreTests
{
    private static DateTimeOffset At(long seconds) => DateTimeOffset.FromUnixTimeSeconds(seconds);

    private static HistoryStore MakeStore(InMemoryClipRepository? repository = null)
    {
        var store = new HistoryStore(repository ?? new InMemoryClipRepository());
        store.Load();
        return store;
    }

    [Fact]
    public void IngestOrdersNewestFirst()
    {
        var store = MakeStore();
        store.Ingest(TextDraft("one"), At(1));
        store.Ingest(TextDraft("two"), At(2));
        Assert.Equal(["two", "one"], store.Rows.Select(r => r.Title));
    }

    [Fact]
    public void DuplicatesBumpInsteadOfInserting()
    {
        var store = MakeStore();
        var first = store.Ingest(TextDraft("same"), At(1));
        store.Ingest(TextDraft("other"), At(2));
        var again = store.Ingest(TextDraft("same", "chrome.exe"), At(3));
        Assert.Equal(first, again);
        Assert.Equal(2, store.Rows.Count);
        var row = store.Row(again)!;
        Assert.Equal(again, store.Rows[0].Id);
        Assert.Equal(2, row.CopyCount);
        Assert.Equal("chrome.exe", row.SourceApp);
        Assert.Equal(At(1), row.FirstCopiedAt);
        Assert.Equal(At(3), row.LastCopiedAt);
    }

    [Fact]
    public void LimitDropsOldestUnpinned()
    {
        var store = MakeStore();
        store.MaxItems = 3;
        for (var i = 0; i < 5; i++) store.Ingest(TextDraft($"item {i}"), At(i));
        Assert.Equal(["item 4", "item 3", "item 2"], store.Rows.Select(r => r.Title));
    }

    [Fact]
    public void PinnedItemsSurviveLimitAndClear()
    {
        var store = MakeStore();
        store.MaxItems = 2;
        var keep = store.Ingest(TextDraft("keep"), At(0));
        store.TogglePin(keep, At(0));
        for (var i = 1; i <= 4; i++) store.Ingest(TextDraft($"item {i}"), At(i));
        Assert.Contains(store.Rows, r => r.Id == keep);
        Assert.Equal(2, store.Rows.Count(r => !r.IsPinned));

        Assert.Equal(2, store.Clear(includingPinned: false));
        Assert.Equal(["keep"], store.Rows.Select(r => r.Title));
        store.Clear(includingPinned: true);
        Assert.Empty(store.Rows);
    }

    [Fact]
    public void DeleteUndoAndPersistence()
    {
        var repository = new InMemoryClipRepository();
        var store = MakeStore(repository);
        var a = store.Ingest(TextDraft("a"), At(1));
        store.TogglePin(a, At(5));
        store.Ingest(TextDraft("b"), At(2));
        var record = store.Delete(a);
        Assert.NotNull(record);
        Assert.Equal(["b"], store.Rows.Select(r => r.Title));
        store.Load();
        Assert.Equal(["b"], store.Rows.Select(r => r.Title));

        var restored = store.Restore(record);
        var row = store.Row(restored)!;
        Assert.Equal("a", row.Title);
        Assert.True(row.IsPinned);
        Assert.Equal(At(1), row.FirstCopiedAt);
        Assert.Equal(["b", "a"], store.Rows.Select(r => r.Title));
        Assert.Equal("a", System.Text.Encoding.UTF8.GetString(store.Contents(restored)[0].Data));
        Assert.Null(store.Delete(Guid.NewGuid()));
    }

    [Fact]
    public void TouchMovesToTop()
    {
        var store = MakeStore();
        var a = store.Ingest(TextDraft("a"), At(1));
        store.Ingest(TextDraft("b"), At(2));
        store.Touch(a, At(3));
        Assert.Equal(a, store.Rows[0].Id);
        Assert.Equal(2, store.Row(a)!.CopyCount);
    }

    [Fact]
    public void ExpiryDropsOldUnpinnedOnly()
    {
        var store = MakeStore();
        store.ExpireAfterDays = 7;
        var now = At(100 * 86_400);
        var old = store.Ingest(TextDraft("old"), now.AddDays(-10));
        var oldPinned = store.Ingest(TextDraft("old pinned"), now.AddDays(-10));
        store.TogglePin(oldPinned, now);
        store.Ingest(TextDraft("fresh"), now.AddDays(-1));
        Assert.Equal(1, store.ExpireOldItems(now));
        Assert.Null(store.Row(old));
        Assert.Equal(["fresh", "old pinned"], store.Rows.Select(r => r.Title).OrderBy(t => t));
        store.ExpireAfterDays = 0;
        Assert.Equal(0, store.ExpireOldItems(now.AddDays(100)));
    }

    [Fact]
    public void LoadRestoresOrderFromTheRepository()
    {
        var repository = new InMemoryClipRepository();
        var store = MakeStore(repository);
        store.Ingest(TextDraft("a"), At(1));
        store.Ingest(TextDraft("b"), At(3));
        store.Ingest(TextDraft("c"), At(2));
        var reloaded = MakeStore(repository);
        Assert.Equal(["b", "c", "a"], reloaded.Rows.Select(r => r.Title));
        Assert.Equal(3, reloaded.Count);
        Assert.Equal(0, reloaded.PinnedCount);
    }

    [Fact]
    public void ClipByHashFindsThePromotedItem()
    {
        var store = MakeStore();
        var draft = TextDraft("promote me");
        var id = store.Ingest(draft, At(1));
        Assert.Equal(id, store.ClipByHash(draft.ContentHash)?.Id);
        Assert.Null(store.ClipByHash("nope"));
    }
}

public class UndoBufferTests
{
    [Fact]
    public void TakeWorksOnlyInsideTheWindowAndOnce()
    {
        var buffer = new UndoBuffer();
        var record = new DeleteRecord(StoredClip.FromDraft(TextDraft("x"), Guid.NewGuid(), T0), []);
        Assert.Null(buffer.Take(T0));
        buffer.Remember(record, T0);
        Assert.True(buffer.HasPending);
        Assert.Same(record, buffer.Take(T0.AddSeconds(3.9)));
        Assert.Null(buffer.Take(T0.AddSeconds(3.9)));
        buffer.Remember(record, T0);
        Assert.Null(buffer.Take(T0.AddSeconds(4.1)));
        Assert.False(buffer.HasPending);
    }
}
